src/lib/ecore/efl_promise2.c

```diff
@@ -0,0 +1,277 @@
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "Ecore.h"
+#include "efl_promise2.h"
+#include "ecore_private.h"
+#include "efl_promise2_priv.h"
+
+struct _Efl_Promise2 {
+   Eina_Free_Cb cancel;
+   Efl_Future2 *future;
+   const void *data;
+};
+
+struct _Efl_Future2 {
+   Efl_Future2_Cb success;
+   Efl_Future2_Cb error;
+   Eina_Free_Cb free;
+   Efl_Promise2 *promise;
+   Efl_Future2 *next;
+   Efl_Future2 *prev;
+   Eina_Value *pending_value;
+   const void *data;
+   Ecore_Event *event;
+   Eina_Bool reject;
+};
+
+static Eina_Mempool *_promise_mp = NULL;
+static Eina_Mempool *_future_mp = NULL;
+static Eina_List *_pending_futures = NULL;
+static int FUTURE_EVENT_ID = -1;
+static Ecore_Event_Handler *_future_event_handler = NULL;
+
+static void
+_efl_future2_dispatch(Efl_Future2 *f, Eina_Value *value, Eina_Bool rejected)
+{
+   while (f)
+     {
+        if (((!rejected && f->success) || (rejected && f->error)))
+          break;
+        Efl_Future2 *n = f->next;
+        Eina_Free_Cb free_cb = f->free;
+        const void *data = f->data;
+        if (n) n->prev = NULL;
+        f->next = NULL;
+        eina_mempool_free(_future_mp, f);
+        if (free_cb) free((void *)data);
+        f = n;
+     }
+
+   if (!f)
+     {
+        eina_value_free(value);
+        return;
+     }
+
+   Eina_Value *next_value = rejected ? f->error((void *)f->data, value) :
+     f->success((void *)f->data, value);
+   if (value != next_value) eina_value_free(value);
+   if (f->free) f->free((void *)f->data);
+
+   Efl_Future2 *next_future = f->next;
+   if (next_future) next_future->prev = NULL;
+   eina_mempool_free(_future_mp, f);
+
+   if (!next_future)
+     {
+        eina_value_free(next_value);
+        return;
+     }
+
+   if (next_value && next_value->type == EINA_VALUE_TYPE_PROMISE2)
+     {
+        Efl_Promise2 *p = eina_value_promise2_steal(next_value);
+        eina_value_free(next_value);
+        f->reject = rejected;
+        p->future = next_future;
+     }
+   else if (((next_value && next_value->type == EINA_VALUE_TYPE_ERROR) || (rejected)))
+     _efl_future2_dispatch(next_future, next_value, EINA_TRUE);
+   else
+     _efl_future2_dispatch(next_future, next_value, EINA_FALSE);
+}
+
+static Eina_Bool
+_efl_future2_handler_cb(void *data EINA_UNUSED,
+                        int type EINA_UNUSED,
+                        void *event)
+{
+   Efl_Future2 *f = event;
+   _efl_future2_dispatch(f, f->pending_value, f->reject);
+   _pending_futures = eina_list_remove(_pending_futures, f);
```

Comment:
_efl_future2_dispatch() is supposed to delete `f`, so I don't think this is necessary.